### ISSUE TITLE
[RFC] Reduce PHP Language Level for Symfony 6 to 7.4

### ISSUE THREAD

**Issue by bobvandevijver** [symfony/symfony#40420]
In #40389 the decision was made to require PHP 8.0 as minimum for the next Symfony major, but not all information given was correct. That's why I propose to lower the minimal language level to 7.4.

**What is wrong with the information in the preceding RFC?**

> PHP 7.4 is currently bundled by the stable and LTS releases of Debian and Ubuntu.

This is _(unfortunately)_ not true for Debian.

**Debian**
Stable Debian (buster/10) only packages 7.3 by default, and the next stable release (bullseye/11, expected somewhere in the second quarter of this year) will only package 7.4: they deemed the transition to 8.0 too large, too breaking and too late. That means that 8.0 (maybe 8.1 by that time?) will only be packaged with the release of Sid, which can be expected in the second quarter of 2023 ([looking at the release history](https://en.wikipedia.org/wiki/Debian_version_history)).

This is relevant Debian thread: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=976811

The PHP packages and Debian versions:
  - [7.3](https://packages.debian.org/search?keywords=php7.3)
  - [7.4](https://packages.debian.org/search?keywords=php7.4)
  - [8.0](https://packages.debian.org/search?keywords=php8.0)

This means that choosing PHP 8.0 for Symfony 6 will close the door for many stable (and thus maintained) Debian users for at least 1.5 years.

**Ubuntu**
The latest LTS of Ubuntu ships with PHP 7.4, so that is true. But that also means that an LTS with PHP 8.0 cannot be expected before 2022.

This will have less impact compared to Debian, but you can still assume that users won't be able to upgrade because they only have access to LTS versions of Ubuntu (of which the installation can postponed by whomever is managing the environment...)

**Conclusion** 
In conclusion, bumping the minimum version to 8.0 might be a bit premature. It will restrict a (possibly large) group of users from upgrading to Symfony 6.0, which has (or at least, not for me) never has been blocked by an LTS OS not having the minimum required PHP version.

Note that all this only holds when you cannot (or do not want to) install a custom package repository such as https://deb.sury.org/. 

As a final note: I'm just opening this to allow the RFC for reconsideration based on this information. Whether or not this changes anything is up to you 😃

**Comment by nicolas-grekas** [comment]
On my side, what matters most to make this decision is the minimum version that 6.4LTS will have to support: because [6.4 will be supported until November 2026](https://symfony.com/releases/6.4), I think going with PHP >= 7.4 would put a too high maintenance burden on the LTS promise. In this regard alone, PHP >= 8.0 is a much better choice for me.

**Comment by derrabus** [comment]
> > PHP 7.4 is currently bundled by the stable and LTS releases of Debian and Ubuntu.
> 
> This is _(unfortunately)_ not true for Debian.

My bad, you are of course right. But at the time Symfony 6.0 is released, Bullseye should‘ve become the stable distribution. Anyway, my point was that if we decide in favor of PHP 8.0, we pick a PHP version that is newer than the bundled version of those distros. And that‘s still true.

Regarding Ubuntu, I would assume that they ship PHP 8.0 with their LTS in April 2022. And that is right on time for Symfony 6.1, which will be the first feature release with new PHP requirements.

Anyhow, PHP 7.4 will be supported by Symfony untill November 2024 (security support until 2025 even). We‘ve got you covered. 🙂 

I personally still think we‘ve made a good decision.

**Comment by jderusse** [comment]
IMHO, comparing LTS distro with non-LTS version of Symfony is not fair.

The next LTS of Symfony is 5.4, released in Nov 2021 (PHP 7.2),
Then 6.4 in Nov 2023. At that time the LTS distro should support natively the minimal requirements.

**Comment by dguhl** [comment]
The Operating System under which PHP will run is out of scope for Symfony's requirement recognition. How Debian and Ubuntu are scheduling PHP distribution can't be a Symfony problem, but would rather be a PHP project's problem. The PHP project would maintain PHP 7.4 for only another year and only in regards to security issues when Symfony 6.0 releases. It would be the last PHP 7-Version maintained but as Symfony 6.0 would require 7, so would 6.4 LTS, which releases long after the end of life for PHP 7.

Debian users who rely on end-of-life PHP versions of their operating system's package maintainers, because "they deemed the transition to 8.0 too large, too breaking and too late", will still have Symfony 5. On Debian, individual Symfony developers would recommend either to compile PHP for the box, to run it from a trustworthy package source like ondrej's launchpad, or to run PHP in a virtual environment like Docker containers, anyway. I don't see the Symfony project issuing official recommendations on that, because, as I said, the operating system is out of scope for the requirement recognition.

**Comment by ghost** [comment]
Hi,

If I can afford it, it's up to distribution to follow the technology and not the other way around. It's like the drivers ... Personally after various tests, on the same machine, the response results were better on Fedora than on debian or ubuntu. +/- 50% of gain ... Then many hosts have more machine under a redHat, CentOs than machine under a deb, for reasons of comfort use, configuration and optimization of a server. It is not at all comparable ...

This to say that distributions follow technologies well, why not others?

regards,

**Comment by nicolas-grekas** [comment]
I'm going to close here because the decision has been made and I think the new argument is not strong enough to change it. Thanks for opening.

**Comment by bobvandevijver** [comment]
For reference, note that this (except for 4.x) breaks with the tradition/rule of thumb of 5 years support outside the PHP support cycle.

Symfony major | maintenance period | min. PHP version | maintenance period | Maintenance diff |
-- | -- | -- | -- | -- |
2.0 – 2.8 | 07/2011 – 09/2019 | 5.3 | 06/2009 – 08/2014 | 5 years
3.0 – 3.4 | 11/2015 – 11/2021 | 5.5 | 06/2013 – 07/2016 | 5 years
4.0 – 4.4 | 11/2017 – 11/2023 | 7.1 | 12/2016 – 12/2019 | 4 years
5.0 – 5.4 | 11/2019 – 11/2025 | 7.2 | 11/2017 – 11/2020 | 5 years
6.0 – 6.4 | 11/2021 – 11/2027 | 7.4 | 11/2019 - 11/2022 | 5 years
6.0 – 6.4 | 11/2021 – 11/2027 | 8.0 | 11/2020 - 11/2023 | 4 years

**Comment by faizanakram99** [comment]
ondrej/ppa supports PHP 8 :smile:  

I am sure most of devs if not all use that PPA over official ones

**Comment by wouterj** [comment]
Something we have learned that is that the LTS releases are really heavy to maintain (e.g. a security fix now must be done on 3.4, to 4.4, to 5.2). If we can lower this burden (even a bit), without putting lots of burden on the community, I think we should do it. Imho, that's what we do by choosing PHP 8.0 vs 7.4.

Anyway, Symfony 6 is released only at the end of this year, and its support will last 6 more years (and 5.4, 4 more years). So there is not much to worry about at this moment :smiley: 

---

*btw, Debian 12 is nicknamed "Bookworm", "Debian Sid" is the nickname for the rolling Debian Unstable release)*

**Comment by bobvandevijver** [comment]
Just to make clear, I'm not against using PHP 8 for Symfony 6: I absolutely agree that it has its advantages over 7.4. I just wanted to share this information and allow everybody to maybe reconsider, as there was an explicit note in the preceding RFC about the bundled PHP versions allowing me to believe it was actually a part of the consideration. 

But, it is now clear that it isn't, and that is perfectly okay with me 👍🏻

-------------
> I am sure most of devs if not all use that PPA over official ones

I don't 😄

> btw, Debian 12 is nicknamed "Bookworm", "Debian Sid" is the nickname for the rolling Debian Unstable release)

Absolutely true, I choose to keep Sid here as the related Debian sources in this PR currently all use that name to avoid confusion.  👍🏻

**Comment by dotdevru** [comment]
Is this joke that framework php min. version is depending on what php version is in debian repo??? You should push devs to use newer php versions instead of non maintanable. Second, how much symfony devs use debian and cannot install php8??? 0.1%?    
I think only the choice for symfony 6.x is php 8 so framework can move towards with language. If u want obsolete technologies stay on symfony 5.x and php 7.x, no?

**Comment by derrabus** [comment]
@dotdevru Please try to maintain a respectful language in your comments. Thank you.

**Comment by dotdevru** [comment]
> @dotdevru Please try to maintain a respectful language in your comments. Thank you.

How this is related to issue itself? What i said wrong? I am trying to explain my opinion, but as i see many ppl there want symfony 6.x be with php 7.x only because debian repo doesnt have php8.0 by default. So, can u explain me pls?

**Comment by dguhl** [comment]
@dotdevru The issue has been decided and closed 5 hours ago.
